File: 61-Rotate-List/solution.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def rotateRight(self, head, k):
        """
        :type head: ListNode
        :type k: int
        :rtype: ListNode
        """
        len, pre, current = 0, None, head
        while current:
            pre, current = current, current.next
            len += 1

        if not len or not k % len:
            return head

        tail = head
        for _ in xrange(len - k % len - 1):
            tail = tail.next

        next, tail.next, pre.next = tail.next, None, head
        return next
        

File: 61-Rotate-List/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        int len = 1;
        ListNode *cur = head;
        
        if (head == NULL) {
            return head;
        }
        
        while (cur->next) {
            len++;
            cur = cur->next;
        }
        
        k = k % len;
        
        if (k == 0) {
            return head;
        }
        
        cur->next = head;  // loop
        
        cur = head;
        
        for (int i = 0; i < len - k - 1; i++) {
            cur = cur->next;
        }
        
        ListNode *const newTail = cur;
        ListNode *const newHead = cur->next;
        
        newTail->next = NULL;
        
        return newHead;
    }
};